### PR TITLE
Recompute validator duties on reorgs across the epoch boundary

### PR DESCRIPTION
**Pull request by potuz**
This PR registers the validator client to the event stream to listen for reorgs. When the reorg crosses the previous epoch boundary it triggers a recomputation of the duties. 

There are two types of reorgs that can happen, in the following if `S'` was the previous slot and `S` is the new head slot

```
     --- S'
   /
-- F
   \
    -----------S
```
Here `depth=S-F` and a new incompatible duty can happen if and only if `epoch(F) < epoch(S)`, this means that
`epoch(S - depth) < epoch (S)`. 

On the other hand if the reorg looks like this

```
   -----------S'
   /
-- F
   \
     --- S
```
That is, we reorg to a previous slot,  `depth=S'-F` and a new duty can happen if and only if `epoch(F) < epoch(S)`, but the client receives `S` and `depth` so it can't compute directly `F`. However `S - depth < S' - depth` so if `epoch(S) = epoch(S')` recomputing in the above case is still good.

There is an extra edge case when the epoch of S' and that of S are different, but in this case we will still recompute the duties. 

- TODO:

[] - unit tests.

### CODE REVIEW

File: validator/client/validator.go [hunk 1]
```diff
@@ -1138,26 +1138,57 @@ func (v *validator) StartEventStream(ctx context.Context, topics []string, event
...
+	}
+	depth, err := strconv.ParseUint(reorg.Depth, 10, 64)
+	if err != nil {
+		log.WithError(err).Error("Failed to parse depth")
+	}
+	if slots.ToEpoch(primitives.Slot(slot-depth)) < slots.ToEpoch(primitives.Slot(slot)) {
+		v.dutiesLock.Lock()
+		v.duties = nil
+		v.dutiesLock.Unlock()
+		if err := v.UpdateDuties(ctx, primitives.Slot(slot)); err != nil {
```
Comment (terencechain) [inline validator/client/validator.go:1172]: `primitives.Slot(slot)` needs to be the epoch start slot right? otherwise `UpdateDuties` would just no-op because of  
```
	if !slots.IsEpochStart(slot) && v.duties != nil {
		// Do nothing if not epoch start AND assignments already exist.
		return nil
	}
```

**Review by terencechain — DISMISSED** [review]

File: validator/client/validator.go [hunk 2]
```diff
@@ -1138,26 +1138,57 @@ func (v *validator) StartEventStream(ctx context.Context, topics []string, event
 	v.validatorClient.StartEventStream(ctx, topics, eventsChannel)
 }
 
-func (v *validator) ProcessEvent(event *eventClient.Event) {
+func (v *validator) processHeadEvent(event *eventClient.Event) {
+	log.Debug("Received head event")
+	head := &structs.HeadEvent{}
+	if err := json.Unmarshal(event.Data, head); err != nil {
+		log.WithError(err).Error("Failed to unmarshal head Event into JSON")
```
Comment (rkapka) [inline validator/client/validator.go:1145]:
```suggestion
		log.WithError(err).Error("Failed to unmarshal head event into JSON")
```

File: validator/client/validator.go [hunk 3]
```diff
@@ -1138,26 +1138,57 @@ func (v *validator) StartEventStream(ctx context.Context, topics []string, event
...
+		log.WithError(err).Error("Failed to parse slot")
+	}
+	v.setHighestSlot(primitives.Slot(uintSlot))
+}
+
+func (v *validator) processReorgEvent(ctx context.Context, event *eventClient.Event) {
+	log.Debug("Received chain reorg event")
+	reorg := &structs.ChainReorgEvent{}
+	if err := json.Unmarshal(event.Data, reorg); err != nil {
+		log.WithError(err).Error("Failed to unmarshal chain reorg Event into JSON")
```
Comment (rkapka) [inline validator/client/validator.go:1158]:
```suggestion
		log.WithError(err).Error("Failed to unmarshal chain reorg event into JSON")
```

File: validator/client/validator.go (hunk 1 above)
Comment (rkapka) [reply validator/client/validator.go:1172]: Setting `v.duties = nil` above takes care of this because there is an `&& v.duties != nil`

File: validator/client/runner.go [hunk 4]
```diff
@@ -297,6 +297,7 @@ func runHealthCheckRoutine(ctx context.Context, v iface.Validator, eventsChan ch
 	log.Info("Starting health check routine for beacon node apis")
 	healthCheckTicker := time.NewTicker(time.Duration(params.BeaconConfig().SecondsPerSlot) * time.Second)
 	tracker := v.HealthTracker()
+	go v.StartEventStream(ctx, event.DefaultEventTopics, eventsChan)
```
Comment (james-prysm) [inline validator/client/runner.go:300]: why do we need this here? I think it should start without this on line 333

File: validator/client/validator.go (hunk 1 above)
Comment (terencechain) [reply validator/client/validator.go:1172]: thanks!

File: validator/client/runner.go (hunk 4 above)
Comment (james-prysm) [reply validator/client/runner.go:300]: potuz answered on slack

File: validator/client/grpc-api/grpc_validator_client.go [hunk 5]
```diff
@@ -214,33 +219,70 @@ func NewGrpcValidatorClient(cc grpc.ClientConnInterface) iface.ValidatorClient {
...
-	if !containsHead {
-		eventsChannel <- &eventClient.Event{
-			EventType: eventClient.EventConnectionError,
-			Data:      []byte(errors.Wrap(client.ErrConnectionIssue, "gRPC only supports the head topic, and head topic was not passed").Error()),
+	c.isEventStreamRunning = true
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Context canceled, stopping event stream")
+			close(eventsChannel)
```
Comment (james-prysm) [inline validator/client/grpc-api/grpc_validator_client.go:236]: i'm not confident this is safe to do. I think if you close this channel here and streamHead is running it will panic if it doesn't do it in sync.

File: validator/client/grpc-api/grpc_validator_client.go [hunk 6]
```diff
@@ -302,6 +344,46 @@ func (c *grpcValidatorClient) StartEventStream(ctx context.Context, topics []str
...
+		eventsChannel <- &eventClient.Event{
+			EventType: eventClient.EventError,
+			Data:      []byte(errors.New("no topics were added").Error()),
+		}
+		return
+	}
+	for _, topic := range supportedRequested {
+		switch topic {
+		case eventClient.EventHead:
+			go c.streamHead(ctx, eventsChannel)
```
Comment (james-prysm) [inline validator/client/grpc-api/grpc_validator_client.go:380]: you might need to use a sync.WaitGroup for closing this channel instead of doing it inside

File: validator/client/grpc-api/grpc_validator_client.go [hunk 7]
```diff
@@ -214,33 +219,70 @@ func NewGrpcValidatorClient(cc grpc.ClientConnInterface) iface.ValidatorClient {
...
+		case <-ctx.Done():
+			log.Info("Context canceled, stopping event stream")
+			close(eventsChannel)
+			c.isEventStreamRunning = false
+			return
+		default:
+			if ctx.Err() != nil {
+				c.isEventStreamRunning = false
+				if errors.Is(ctx.Err(), context.Canceled) {
+					eventsChannel <- &eventClient.Event{
```
Comment (james-prysm) [inline validator/client/grpc-api/grpc_validator_client.go:243]: another concern is about the buffer size here, I think it's only 1 `eventsChan := make(chan *event.Event, 1)`
so I think the channel should only take 1 event at a time but if this sends it both from different routines without the reader reading it does it error being filled up? not totally confident in this answer, might need to double check with @kasey .

**Comment by potuz** [comment]
Closed in favor of #15142  just to make it easier to diagnose other CLs VCs since they use that approach mostly, but I would rather use this one with the reorg event instead.
